lib/services/identity.go

```diff
@@ -207,6 +207,21 @@ type Identity interface {
 	// GetUserTokenSecrets returns a user token secrets.
 	GetUserTokenSecrets(ctx context.Context, tokenID string) (types.UserTokenSecrets, error)
 
+	// UpsertRecoveryCodes upserts a user's new recovery codes.
+	UpsertRecoveryCodes(ctx context.Context, user string, recovery types.RecoveryCodesV1) error
+
+	// GetRecoveryCodes gets a user's recovery codes.
+	GetRecoveryCodes(ctx context.Context, user string) (*types.RecoveryCodesV1, error)
+
+	// CreateUserRecoveryAttempt logs user recovery attempt.
+	CreateUserRecoveryAttempt(ctx context.Context, user string, attempt types.RecoveryAttempt) error
```

Comment:
Suggestion: pass structs by pointer.

```suggestion
	CreateUserRecoveryAttempt(ctx context.Context, user string, attempt *types.RecoveryAttempt) error
```